File: May Challenge 2019/REDONE_2.cpp
#include<bits/stdc++.h>
using namespace std;

#define lli long long int

int main()
{
    int test; cin >> test;

    lli arr[1000010];
    for(lli i=0; i<1000010; i++)
        arr[i] = 0;
    arr[0] = 1;

    lli mod = 1000000007; 
    while(test--)
    {
        lli n; cin >> n;
        if(arr[n-1] != 0)
            cout << arr[n-1]%mod << endl;
        else
        {
            for(lli i=1; i<n; i++)
            {
                arr[i] = (i+1 + arr[i-1] + (i+1)*arr[i-1])%mod;
            }
            cout << arr[n-1]%mod << endl;
        }
    }
}

File: May Challenge 2019/REDONE_2.py
mod = 1000000007

arr = [0 for i in range(1000000)]
arr[0] = 1

for _ in range(int(input())):
    n = int(input())

    if arr[n-1] != 0:
        print(arr[n-1]%mod)
    else:
        for i in range(1, n):
            arr[i] = (i+1 + arr[i-1] + (i+1)*arr[i-1])%mod
            print(str(i) + " --> " + str(arr[i]%mod))
        print(arr[n-1]%mod)